jaxlib/mosaic/gpu/custom_call.cc

```diff
@@ -137,26 +138,60 @@ absl::Status RunCUDATool(const char* tool,
   std::string tool_path(cuda_path_ptr);
   tool_path += "/bin/";
   tool_path += tool;
+  int stdout_pipe[2];
   pid_t child_pid;
   posix_spawn_file_actions_t file_actions;
   if (posix_spawn_file_actions_init(&file_actions)) {
     return absl::InternalError("Failed to initialize spawn file actions");
   }
+  if (pipe(stdout_pipe) == -1) {
+    return absl::InternalError("Failed to set up pipe");
+  }
+  // close read end in child
+  if (posix_spawn_file_actions_addclose(&file_actions, stdout_pipe[0])) {
+    return absl::InternalError("Failed to close read end of the pipe in child");
+  }
+  if (posix_spawn_file_actions_adddup2(&file_actions, stdout_pipe[1],
+                                       STDOUT_FILENO)) {
+    return absl::InternalError("Failed to redirect stdout to pipe");
+  }
   if (posix_spawn_file_actions_adddup2(&file_actions, STDOUT_FILENO,
                                        STDERR_FILENO)) {
-    return absl::InternalError("Failed to set up spawn file actions");
+    return absl::InternalError("Failed to redirect stderr to stdout");
   }
   // execv is guaranteed by POSIX to not modify the args (other than
   // replacing the whole process image), so the const_cast is valid.
-  if (posix_spawn(&child_pid, tool_path.c_str(), &file_actions, nullptr,
-                  const_cast<char* const*>(args.data()), environ)) {
-    return absl::InternalError("Process spawn failed");
+  if (int status =
+          posix_spawn(&child_pid, tool_path.c_str(), &file_actions, nullptr,
+                      const_cast<char *const *>(args.data()), environ)) {
+    return absl::InternalError(
+        absl::StrCat("Process spawn failed: ", strerror(status)));
+  }
+  // close write end in parent
+  if (close(stdout_pipe[1]) == -1) {
+    return absl::InternalError(
+        absl::StrCat("Failed to close write end of pipe in parent process: ",
+                     strerror(errno)));
+  }
+  std::string stdout;
+  char buf[1024];
+  while (int status = read(stdout_pipe[0], buf, sizeof buf)) {
+    if (status == -1) {
+      return absl::InternalError(
+          absl::StrCat("Failed to read from pipe: ", strerror(errno)));
+    }
+    stdout += buf;
+  }
+  if (close(stdout_pipe[0]) == -1) {
```

Comment:
Please use absl::Cleanup to close the pipes. This, or the close above might not get executed if something else fails along the way

Reply:
Done.